myia/dtype.py

```diff
@@ -0,0 +1,188 @@
+
+    def __call__(self, *args, **kwargs):
+        """Overloaded to change the semantics of class instantiation."""
+        if hasattr(self, '_parse_args'):
+            args = self._parse_args(args, kwargs)
+        else:
+            assert len(kwargs) == 0
+        key = (self, args)
+        if key not in self._instances:
+            obj = super().__call__(*args)
+            self._instances[key] = obj
+        return self._instances[key]
+
+
+class Type(metaclass=TypeMeta):
+    """Base class for all Types."""
+
+    _root = True
+    _fields: DictT[str, Any]
+
+    def __new__(self, *args):
+        """Disable instantiation of the base type."""
+        raise RuntimeError("Can't instantiate Type")
+
+    def __eq__(self, other):
+        return self is other
+
+    def __hash__(self):
+        return id(self)
+
+    def __repr__(self) -> str:
+        name = type(self).__name__
+        args = ', '.join(repr(getattr(self, p)) for p in self._fields.keys())
+        return f"{name}({args})"
+
+
+class Bool(Type):
+    """Boolean values."""
+
+
+class Number(Type):
+    """Numerical values."""
+
+    bits: int
+    _valid_bits: TupleT[int, ...] = ()
+
+    @classmethod
+    def _parse_args(cls, args, kwargs):
+        assert cls is not Number
+        assert len(kwargs) == 0
+        assert args[0] in cls._valid_bits
+        return args
+
+
+class Float(Number):
+    """Represents float values."""
+
+    _valid_bits = (16, 32, 64)
+
+
+class Int(Number):
+    """Represents signed integer values."""
+
+    _valid_bits = (8, 16, 32, 64)
+
+
+class UInt(Number):
+    """Represents unsigned integer values."""
+
+    _valid_bits = (8, 16, 32, 64)
+
+
+class List(Type):
+    """Represents a set of ordered values with the same type."""
+
+    element_type: Type
+
+
+class Struct(Type):
+    """Represents a set of named fields with their own types."""
+
+    elements: DictT[str, Type]
+
+    @classmethod
+    def _parse_args(cls, args, kwargs):
+        if len(kwargs) != 0:
+            assert len(args) == 0
+            elems = kwargs
+        else:
+            assert len(args) == 1
+            elems = dict(args[0])
+        items = ((k, elems[k]) for k in sorted(elems.keys()))
+        return (tuple(items),)
+
+    def __new__(cls, elements: KeysT) -> None:
+        """Convert input to a dict."""
+        return tuple.__new__(cls, (dict(elements),))  # type: ignore
+
+    def __getattr__(self, attr):
+        if attr in self.elements:
+            return self.elements[attr]
+        raise AttributeError
+
+
+class Tuple(Type):
+    """Represents a set of ordered values with independant types."""
+
+    elements: TupleT[Type, ...]
+
+    @classmethod
+    def _parse_args(cls, args, kwargs):
+        assert len(kwargs) == 0
+        if len(args) == 1 and isinstance(args[0], collections.Iterable):
+            return (tuple(args[0]),)
+        else:
+            return (args,)
+
+
+class Callable(Type):
```

Comment:
`Callable` feels like Python jargon, I'd prefer to call this `Function`.

Reply:
I'm ok with that.